[flang] Use precompiled parsing headers

mrkajetanp opened the pull request:
Most of the high memory usage and compilation time in the frontend units is due to including large parsing headers.
This commit moves out several of the largest parsing headers into a new precompiled header linked to flangFrontend.
The new compilation metrics for FrontendActions.cpp are as follows:

User time (seconds): 38.40
System time (seconds): 2.00
Maximum resident set size (kbytes): 2710964 (2.58 GB) (vs 3.78 GB)

ParserActions.cpp:

User time (seconds): 69.37
System time (seconds): 1.81
Maximum resident set size (kbytes): 2599456 (2.47 GB) (vs 4 GB)

Alongside the new precompiled header compilation unit

User time (seconds): 41.61
System time (seconds): 2.72
Maximum resident set size (kbytes): 3107644 (2.96 GB)

mrkajetanp committed:
[flang] Use precompiled parsing headers

Most of the high memory usage and compilation time in the frontend units
is due to including large parsing headers.
This commit moves out several of the largest parsing headers into a new
precompiled header linked to flangFrontend.
The new compilation metrics for FrontendActions.cpp are as follows:

User time (seconds): 38.40
System time (seconds): 2.00
Maximum resident set size (kbytes): 2710964 (2.58 GB) (vs 3.78 GB)

ParserActions.cpp:

User time (seconds): 69.37
System time (seconds): 1.81
Maximum resident set size (kbytes): 2599456 (2.47 GB) (vs 4 GB)

Alongside the new precompiled header compilation unit

User time (seconds): 41.61
System time (seconds): 2.72
Maximum resident set size (kbytes): 3107644 (2.96 GB)

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

tblah commented on flang/lib/Frontend/CMakeLists.txt:
```diff
@@ -72,3 +72,11 @@ add_flang_library(flangFrontend
   clangDriver
 )
+
+target_precompile_headers(flangFrontend PRIVATE
+  [["flang/Parser/parsing.h"]]
+  [["flang/Parser/parse-tree.h"]]
+  [["flang/Parser/dump-parse-tree.h"]]
+  [["flang/Lower/PFTBuilder.h"]]
+  [["flang/Lower/Bridge.h"]]
+)
```
Is there a way these could be kept in the part of flang the headers are from and then re-used here?
e.g. the first three would be precompiled in flang/lib/Parser/CMakeLists.txt

DavidTruby replied:
In general it's only safe to have one precompiled header per TU, so that's all CMake (and most compilers) support. CMake isn't so much compiling the "Lower" headers _here_ though, conceptually (at least, not any more than including them in other files is).

CMake does something like the following
1. creates a `flangFrontend_headers.h` file that just contains `#include "flang/Parser/parsing.h` etc
2. asks the compiler to compile that single header into `flangFrontend_headers.pch` making sure to use exactly the same flags as will be used for each object file in the target
3. Includes that flangFrontend_headers.pch file in all the TUs in that target.

So there's nothing visible to outside this process that indicates that the "Lower" headers are being "built" here really.

As a result if you wanted to share a single `.pch` for the entire project, you have to do something like
```
target_precompile_headers(flangLowering REUSE-FROM flangFrontend)
```
because even if you list the exact same files, CMake will follow the process above by default and not re-use them, because each file isn't being treated individually but as part of a single `.pch` file for the whole target.

tblah replied:
What I am interested in here is whether we can design something that scales to using precompiled headers in other very slow translation units (e.g. flang/lib/Lower/*)

mrkajetanp replied:
We could have Parser/, Lower/, Frontend/ etc. each have their own set of precompiled headers. There would be overlap between those for sure (i.e. we'd be precompiling the parse tree for each of them), but it could still improve the overall compile time.

DavidTruby replied:
That's generally what you would do, since the set of precompiled headers you want is probably different for each target (even if there is some overlap). This does lead to _some_ duplication of work, but certainly no more than there is now where we are compiling these headers not only in every target, but in every _TU_!

tblah replied:
It's up to you if you want to do it in this PR or a follow up

DavidTruby commented on flang/lib/Frontend/CompilerInstance.cpp:
```diff
@@ -13,7 +13,6 @@
 #include "flang/Frontend/CompilerInstance.h"
 #include "flang/Frontend/CompilerInvocation.h"
 #include "flang/Frontend/TextDiagnosticPrinter.h"
-#include "flang/Parser/parsing.h"
```
I don't think you need/want to remove these headers for this to work, and there's a couple of disadvantages to doing so: for example, clangd/other editor support might break if the right "real" includes aren't there.

These includes will get immediately skipped over in favour of the `.pch` includes anyway, because of the guard macros (`#ifndef FLANG_FRONTEND_COMPILERINSTANCE_H` etc) and the `.pch` being included first. So I think it's best to leave them in.

mrkajetanp replied:
FWIW clangd and other LSPs can handle this (at least mine seems fine) because they just pull from compile_commands.json. But you're right on the wider point, the the build system actually handles leaving this just fine. I'll fix the PR in a moment :)

Meinersbur replied:
What happens with [`CMAKE_DISABLE_PRECOMPILE_HEADERS=ON`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_PRECOMPILE_HEADERS.html#variable:CMAKE_DISABLE_PRECOMPILE_HEADERS)?

Also, what happens when compiling with gcc (or any other compiler), then using clangd using `compile_commands.json`?

DavidTruby reviewed (commented):
This looks good to me, it's great to see some improvement in memory usage being worked on here!

Did this have much impact on compile time as well? I see the memory usage clearly coming down in your commit message but only the new compile time numbers and not the old ones?

I'll give it a try myself when I have a sec and see how it works here.
Thanks!

mrkajetanp commented:
> Did this have much impact on compile time as well? I see the memory usage clearly coming down in your commit message but only the new compile time numbers and not the old ones?

It did, I just skipped over those for brevity (and because they're very similar to those included in the previous flang commit for splitting the TUs). Here are the last measurements I got:

FrontendActions.cpp:
User time (seconds): 38.40 (vs 69.35)
System time (seconds): 2.00 (vs 3.56)

ParserActions.cpp:
User time (seconds): 69.37 (vs 109.19)
System time (seconds): 1.81 (vs 3.80)

Precompiled header:
User time (seconds): 41.61
System time (seconds): 2.72

kparzysz commented:
I'm not in favor of this, I think this is way too fragile.  We should favor methods that are more self-explanatory (and thus harder to break unknowingly).

mrkajetanp committed:
[flang] Add back precompiled header includes

The build system can handle prioritising pre-compiled headers over
explicit includes if they are available. There is thus no need to remove
the explicit includes.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp commented:
> I'm not in favor of this, I think this is way too fragile. We should favor methods that are more self-explanatory (and thus harder to break unknowingly).

Hopefully David's suggestion should assuage the concerns here - this doesn't actually require removing any of the already existing includes. CMake can handle this completely transparently.

mrkajetanp commented:
> I often try to compile individual files by hand when I'm dealing with compilation issues to save time. Fixing these issues may require changing headers

I gave it try, as it stands now this still works completely fine. You can just drop the `pch` includes from the compile command for a given TU and it'll pull in the headers normally as before. Slower to compile but won't break.
Clang is pretty good with this too, so even if you forget to drop it but change the header that was being precompiled clang will let you know that the precompiled header needs to be rebuilt because the source has changed.

DavidTruby commented:
> Clang is pretty good with this too, so even if you forget to drop it but change the header that was being precompiled clang will let you know that the precompiled header needs to be rebuilt because the source has changed.

FWIW gcc will do the same, and will also warn you if your _flags_ have changed and the pch uses old ones (I think clang does the latter as well, but I haven't checked!)

DavidTruby reviewed (approved):
This LGTM once the other reviewers are happy!

I think you should just post the other PCH changes (for e.g. Parser and Lower) in a follow up, because we can let this one work its way through the build bots first and see if there are any issues in the simple case before expanding it.

mrkajetanp commented:
> This LGTM once the other reviewers are happy!
> 
> I think you should just post the other PCH changes (for e.g. Parser and Lower) in a follow up, because we can let this one work its way through the build bots first and see if there are any issues in the simple case before expanding it.

Agreed, sounds sensible :)

tblah commented:
@kparzysz are you happy to proceed [with the `#include`s not removed](https://github.com/llvm/llvm-project/pull/130600/commits/f9662a4eea7749b6e58ffbc085ea630ff9ab6b39) or would you still prefer not to see this change?

mgorny commented:
Thanks! I'm not 100% sure whether it worked, since the build OOM-ed hard for me with `-j3` anyway, but it didn't OOM on `lib/Frontend`, so perhaps it did.

For me, the worst offender is `lib/Lower`, with `lib/Semantics` going next, and then probably `lib/Evaluate` and `lib/Optimizer/Transforms` (these three OOM with `-j4`).

Meinersbur commented:
There is a problem in that it breaks some ccache builds:
 * https://lab.llvm.org/staging/#/builders/206/builds/174

In any case, it renders ccache ineffective (without additional accomodations):
 * https://ccache.dev/manual/latest.html#_precompiled_headers

mrkajetanp commented:
Ah that's annoying, thanks for pointing it out! Is it just the MSVC builds? I'm wondering if there's some way I could reproduce this ad-hoc and I don't have access to a Windows machine. I'm doing ccache builds on AArch64 Linux and it's working fine.
Regardless, we're already using `-include-pch` so I think we just need to make sure we use `-fno-pch-timestamp`and set `CCACHE_SLOPPINESS` appropriately?

mrkajetanp commented:
How about this?
https://github.com/llvm/llvm-project/pull/131397
